internal/contentmanager/contentmanager.go

```diff
@@ -0,0 +1,203 @@
+		listKind := obj.GetObjectKind().GroupVersionKind().Kind + "List"
+
+		// If the Kind or Version is not set in an object's GroupVersionKind
+		// attempting to add it to the runtime.Scheme will result in a panic.
+		// To avoid panics, we are doing the validation and returning early
+		// with an error if any objects are provided with a missing Kind or Version
+		// field
+		if gvk.Kind == "" {
+			return nil, fmt.Errorf(
+				"adding %s to scheme; object Kind is not defined",
+				obj.GetName(),
+			)
+		}
+
+		if gvk.Version == "" {
+			return nil, fmt.Errorf(
+				"adding %s to scheme; object Version is not defined",
+				obj.GetName(),
+			)
+		}
+
+		if !scheme.Recognizes(gvk) {
+			// Since we can't have a mapping to every possible Go type in existence
+			// based on the GVK we need to use the unstructured types for mapping
+			u := &unstructured.Unstructured{}
+			u.SetGroupVersionKind(gvk)
+			ul := &unstructured.UnstructuredList{}
+			ul.SetGroupVersionKind(gvk.GroupVersion().WithKind(listKind))
+
+			scheme.AddKnownTypeWithName(gvk, u)
+			scheme.AddKnownTypeWithName(gvk.GroupVersion().WithKind(listKind), ul)
+			// Adding the common meta schemas to the scheme for the GroupVersion
+			// is necessary to ensure the scheme is aware of the different operations
+			// that can be performed against the resources in this GroupVersion
+			metav1.AddToGroupVersion(scheme, gvk.GroupVersion())
+		}
+	}
+
+	return scheme, nil
+}
+
+// ManageContent configures a controller-runtime cache.Cache and establishes watches for the provided resources.
+// It utilizes the provided ClusterExtension to set a DefaultLabelSelector on the cache.Cache
+// to ensure it is only caching and reacting to content that belongs to the ClusterExtension.
+// For each client.Object provided, a new source.Kind is created and used in a call to the Watch() method
+// of the provided controller.Controller to establish new watches for the managed resources.
+func (i *instance) ManageContent(ctx context.Context, ctrl controller.Controller, ce *v1alpha1.ClusterExtension, objs []client.Object) error {
+	// Return nil if the objs slice is empty before any further action
+	if len(objs) == 0 || ce == nil || ctrl == nil {
+		return nil
+	}
+
+	cfg, err := i.rcm(ctx, ce, i.baseCfg)
+	if err != nil {
+		return fmt.Errorf("getting rest.Config for ClusterExtension %q: %w", ce.Name, err)
+	}
+
+	scheme, err := buildScheme(objs)
+	if err != nil {
+		return fmt.Errorf(
+			"building scheme for %s; %w",
+			ce.GetName(),
+			err,
+		)
+	}
+
+	tgtLabels := labels.Set{
+		oclabels.OwnerKindKey: v1alpha1.ClusterExtensionKind,
+		oclabels.OwnerNameKey: ce.GetName(),
+	}
+
+	c, err := cache.New(cfg, cache.Options{
+		Scheme:               scheme,
+		DefaultLabelSelector: tgtLabels.AsSelector(),
+	})
+	if err != nil {
+		return fmt.Errorf("creating cache for ClusterExtension %q: %w", ce.Name, err)
+	}
+
+	for _, obj := range objs {
+		err = ctrl.Watch(
+			source.Kind(
```

Comment:
I know I asked similar questions to what I'm about to ask in the Brief/RFC, but now with a new slant.

We're cancelling the cache after we're done with it, but the `source.Kind`s stick around forever? Is this a memory leak?

Reply:
Actually no, this wouldn't be a memory leak. I just dug a bit more into the controller-runtime internal logic and it looks like sources are temporary:
- When calling `ctrl.Watch()` the source is only stored if the controller has not yet been started, otherwise it's Start function is called and no reference is stored
    - Ref: https://github.com/kubernetes-sigs/controller-runtime/blob/a0c9fd9d3f310f48155ce985366b21914675fbea/pkg/internal/controller/controller.go#L117-L131
- When calling `ctrl.Start()` all sources stored to be started are started and then no longer stored as references so they can be garbage collected when no longer in use
    - Ref: https://github.com/kubernetes-sigs/controller-runtime/blob/a0c9fd9d3f310f48155ce985366b21914675fbea/pkg/internal/controller/controller.go#L171-L212

Reply:
That looks promising. Thanks!

I followed that path of `source.Kind.Start()` as well. It looks like that function:
1. Get's an informer from the cache.
2. Adds the specified event hander to it, but notably does _not_ capture the registration struct, which can be used to remove the event handler from the informer later.  Maybe this isn't important to us though because we are probably not adding/removing event handlers from informers. We would be adding/removing entire informers.
3. We stop the entire cache when we're done with it, which should stop all of the informers and event handlers, and free up any referenced resources.